admiral/pkg/clusters/serviceentry.go

```diff
@@ -554,8 +572,26 @@ func generateServiceEntry(admiralCache *AdmiralCache, meshPorts map[string]uint3
 	}
 	seEndpoint := makeRemoteEndpointForServiceEntry(endpointAddress,
 		locality, finalProtocol, port)
-	tmpSe.Endpoints = append(tmpSe.Endpoints, seEndpoint)
+
+	// if the action is deleting an endpoint from service entry, loop through the list and delete matching ones
+	if event == common.Add {
+		tmpSe.Endpoints = append(tmpSe.Endpoints, seEndpoint)
+	} else if event == common.Delete {
+		for i, exitingEndpoint := range tmpSe.Endpoints {
+			if reflect.DeepEqual(exitingEndpoint, seEndpoint) {
+				tmpSe.Endpoints = RemoveIndex(tmpSe.Endpoints, i)
```

Comment:
Is this not causing index problems as you iterate and remove elements from the same list? why not creating another with filtered endpoints and swap it at the end. Definitely add a test case for this.

Reply:
yep, I made a mistake here I think it might caused the problem you said